presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +611,202 @@ private static NestedField nestedField(String name)
+                            .orElse(IntStream.range(0, call.getArguments().size()).boxed().collect(toImmutableSet()));
+
+                    // All the lambda subfields collected in outer functions relate only to the arguments of the function specified in
+                    // functionDescriptor.argumentIndicesContainingMapOrArray.
+                    Map<Integer, Set<Subfield>> lambdaSubfieldsFromOuterFunctions = argumentIndicesContainingMapOrArray.stream()
+                            .collect(toImmutableMap(callArgumentIndex -> callArgumentIndex, unused -> ImmutableSet.copyOf(context.getLambdaSubfields())));
+
+                    // If the function accepts lambdas, add all the lambda subfields from each lambda.
+                    Map<Integer, Set<Subfield>> lambdaSubfieldsFromCurrentFunction = ImmutableMap.of();
+                    for (LambdaDescriptor lambdaDescriptor : functionDescriptor.getLambdaDescriptors()) {
+                        Optional<Map<Integer, Set<Subfield>>> lambdaSubfields = collectLambdaSubfields(call, lambdaDescriptor);
+                        if (!lambdaSubfields.isPresent()) {
+                            context.giveUpOnCollectingLambdaSubfields();
+                            call.getArguments().forEach(argument -> argument.accept(this, context));
+                            return null;
+                        }
+                        lambdaSubfieldsFromCurrentFunction = merge(lambdaSubfieldsFromCurrentFunction, lambdaSubfields.get());
+                    }
+
+                    Map<Integer, Set<Subfield>> lambdaSubfields = merge(lambdaSubfieldsFromOuterFunctions, lambdaSubfieldsFromCurrentFunction);
+
+                    lambdaSubfields = addNoSubfieldIfNoAccessedSubfieldsFound(call, lambdaSubfields);
+
+                    // We need to continue visiting the function arguments and collect all lambda subfields in inner function calls as well as non-lambda subfields in all
+                    // function arguments. Once reached the leaf node, we will try to prune the subfields of the input field, subscript, or subfield.
+                    for (int callArgumentIndex = 0; callArgumentIndex < call.getArguments().size(); callArgumentIndex++) {
+                        // Since context is global during the traversal of all the nodes in expression tree, we need to pass lambda subfields only to those  function
+                        // arguments that they relate to.
+                        if (lambdaSubfields.containsKey(callArgumentIndex)) {
+                            context.setLambdaSubfields(lambdaSubfields.get(callArgumentIndex));
+                        }
+                        else {
+                            context.setLambdaSubfields(Context.allSubfieldsOfArrayElementOrMapValue);
+                        }
+                        call.getArguments().get(callArgumentIndex).accept(this, context);
```

Comment:
If we have lambda arguments, do we visit them multiple times? collectLambdaSubfields() also visits expression arguments if they are lambda

Reply:
Yes. In collectLambdaSubfields, lambda arguments traversed with isolated sub-context for the purpose of extracting lambda subfields only. Here, we need to traverse them with the full context in order to collect all non-lambda subfields.